src/client/Client.cc

```diff
@@ -6848,6 +6876,19 @@ int Client::_lookup(Inode *dir, const string& dname, int mask, InodeRef *target,
     } else {
       ldout(cct, 20) << " no cap on " << dn->inode->vino() << dendl;
     }
+
+    // In rare case during the rename if another thread tries to
+    // lookup the dst dentry, it may get an inconsistent result
+    // that both src dentry and dst dentry will link to the same
+    // inode at the same time.
+    // Will wait the rename to finish and try it again.
+    if (!is_rename && dn->is_renaming) {
+      ldout(cct, 1) << __func__ << " dir " << *dir
+                    << " rename is on the way, will wait for dn '"
+                    << dname << "'" << dendl;
+      wait_on_list(waiting_for_rename);
```

Comment:
What protects this condition and wait? It looks like the client_lock. Will that still be the case after your work to break up the client_lock? I assume we'll want to avoid taking that lock in lookup if we can help it...

Reply:
No, currently I am using the `client_lock` for this and queue the waiters in a global list in the `Client` instead of each `Dentry`.

And after the inode lock feature is finished or in that patch set I will switch this to inode lock instead.